File: web/src/tabs.test.js
import { describe, it, expect } from "vitest";
import { normalizeTabs, resolveTabs, presetTabs } from "./tabs";

describe("normalizeTabs", () => {
  it("empty is empty", () => expect(normalizeTabs([], true)).toEqual([]));
  it("canonical order + dedupe", () =>
    expect(normalizeTabs(["logs", "files", "files", "assistant"], true)).toEqual(["assistant", "files", "logs"]));
  it("drops unknowns", () => expect(normalizeTabs(["files", "bogus", "settings"], true)).toEqual(["files"]));
  it("forces files when no primary", () =>
    expect(normalizeTabs(["terminal", "logs"], true)).toEqual(["files", "terminal", "logs"]));
  it("assistant alone allowed when enabled", () => expect(normalizeTabs(["assistant"], true)).toEqual(["assistant"]));
  it("drops assistant + forces files when disabled", () =>
    expect(normalizeTabs(["assistant", "terminal"], false)).toEqual(["files", "terminal"]));
});

describe("resolveTabs", () => {
  it("app override wins", () =>
    expect(resolveTabs("assistant", "files,logs", true)).toEqual(["assistant"]));
  it("falls back to profile default", () =>
    expect(resolveTabs("", "files,logs", true)).toEqual(["files", "logs"]));
  it("falls back to built-in (all) when nothing set", () =>
    expect(resolveTabs("", "", true)).toEqual(["assistant", "files", "terminal", "logs"]));
  it("built-in without assistant when disabled", () =>
    expect(resolveTabs("", "", false)).toEqual(["files", "terminal", "logs"]));
});

describe("presetTabs", () => {
  it("novice is assistant only", () => expect(presetTabs("novice", true)).toEqual(["assistant"]));
  it("novice with no assistant is files", () => expect(presetTabs("novice", false)).toEqual(["files"]));
  it("expert is everything", () =>
    expect(presetTabs("expert", true)).toEqual(["assistant", "files", "terminal", "logs"]));
});

File: control/tabs_test.go
package control

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNormalizeTabs(t *testing.T) {
	t.Parallel()
	tests := []struct {
		name             string
		in               string
		assistantEnabled bool
		want             string
	}{
		{"empty stays empty (no override)", "", true, ""},
		{"canonical order + dedupe", "logs,files,files,assistant", true, "assistant,files,logs"},
		{"unknown keys dropped", "files,bogus,settings", true, "files"},
		{"neither assistant nor files forces files", "terminal,logs", true, "files,terminal,logs"},
		{"assistant alone is allowed when enabled", "assistant", true, "assistant"},
		{"assistant dropped and files forced when disabled", "assistant,terminal", false, "files,terminal"},
		{"assistant-only becomes files when disabled", "assistant", false, "files"},
		{"whitespace tolerated", " files , logs ", true, "files,logs"},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.want, normalizeTabs(tc.in, tc.assistantEnabled))
		})
	}
}
